### PR TITLE
Adding C++ auth metadata processor.

### PR DESCRIPTION
**Pull request by jboeuf**
- We always do the processing asynchronously but maintain a synchronous
  API for the implementor of the processor.
- there are a lot of string copies right now. Having a StringPiece
  object in grpc++ would really help with that (as we would use it for
  C++ metadata).
- Please review the API carefully and if you're happy with it, I'll
  proceed with tests.

### CODE REVIEW

**Comment by jboeuf** [comment]
@ctiller may want to have a look as well (string copy issues).

File: src/cpp/server/insecure_server_credentials.cc [hunk 1]
```diff
@@ -43,6 +43,8 @@ class InsecureServerCredentialsImpl GRPC_FINAL : public ServerCredentials {
                       grpc_server* server) GRPC_OVERRIDE {
     return grpc_server_add_insecure_http2_port(server, addr.c_str());
   }
+  void SetAuthMetadataProcessor(
+      const std::shared_ptr<AuthMetadataProcessor>& processor) GRPC_OVERRIDE {}
```
Comment (jboeuf) [inline src/cpp/server/insecure_server_credentials.cc]: maybe we should assert fail here...

File: include/grpc++/auth_metadata_processor.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  // If this method returns true, the Process function will be scheduled in
+  // a different thread as the one processing the call.
+  virtual bool IsBlocking() const { return true; }
+
+  // context is read/write: it contains the properties of the channel peer and
+  // it is the job of the Process method to augment it with properties derived
+  // from the passed-in auth_metadata.
+  virtual bool Process(
+      std::multimap<grpc::string, grpc::string>& auth_metadata,
```
Comment (yang-g) [inline include/grpc++/auth_metadata_processor.h]: const?

File: include/grpc++/auth_metadata_processor.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  virtual ~AuthMetadataProcessor() {}
+
+  // If this method returns true, the Process function will be scheduled in
+  // a different thread as the one processing the call.
+  virtual bool IsBlocking() const { return true; }
+
+  // context is read/write: it contains the properties of the channel peer and
+  // it is the job of the Process method to augment it with properties derived
+  // from the passed-in auth_metadata.
+  virtual bool Process(
```
Comment (yang-g) [inline include/grpc++/auth_metadata_processor.h]: Document return value and consumed_auth_metadata?

File: include/grpc++/auth_context.h [hunk 4]
```diff
@@ -87,6 +87,11 @@ class AuthContext {
   // Iteration over all the properties.
   virtual AuthPropertyIterator begin() const = 0;
   virtual AuthPropertyIterator end() const = 0;
+
+  // Mutation functions: should only be used by an AuthMetadataProcessor.
+  virtual void AddProperty(const grpc::string& key,
```
Comment (yang-g) [inline include/grpc++/auth_context.h]: So there is no need of removing/replacing a metadata?

File: src/cpp/server/insecure_server_credentials.cc (hunk 1 above)
Comment (yang-g) [inline src/cpp/server/insecure_server_credentials.cc]: +1

File: include/grpc++/server_credentials.h [hunk 5]
```diff
@@ -54,6 +55,10 @@ class ServerCredentials {
 
   virtual int AddPortToServer(const grpc::string& addr,
                               grpc_server* server) = 0;
+
+  // Has to be called before the server is started.
```
Comment (yang-g) [inline include/grpc++/server_credentials.h]: Maybe add a comment saying should only be called once or the last one counts or something?

File: src/cpp/server/insecure_server_credentials.cc (hunk 1 above)
Comment (jboeuf) [inline src/cpp/server/insecure_server_credentials.cc]: Done.

File: include/grpc++/auth_metadata_processor.h (hunk 2 above)
Comment (jboeuf) [inline include/grpc++/auth_metadata_processor.h]: of course. Sorry...

File: include/grpc++/auth_metadata_processor.h (hunk 3 above)
Comment (jboeuf) [inline include/grpc++/auth_metadata_processor.h]: Good point. Done.

File: include/grpc++/auth_context.h (hunk 4 above)
Comment (jboeuf) [inline include/grpc++/auth_context.h]: You mean a property right?
So, yes, this is the intention. You cannot remove/replace properties from the passed-in auth context (which basically comes from the channel and is chained at the lowest layer). This is Add only.

File: include/grpc++/server_credentials.h (hunk 5 above)
Comment (jboeuf) [inline include/grpc++/server_credentials.h]: Good point. Done.

**Comment by yang-g** [comment]
API LGTM

**Comment by jboeuf** [comment]
For this one, we're missing the string_ref class that I will hopefully send for review today (sorry about the delay).

**Comment by jboeuf** [comment]
@yang-g. The implementation is complete. Can you have a look?

File: include/grpc++/auth_metadata_processor.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef GRPCXX_AUTH_METADATA_PROCESSOR_H_
+#define GRPCXX_AUTH_METADATA_PROCESSOR_H_
```
Comment (yang-g) [inline include/grpc++/auth_metadata_processor.h]: This seems to belong to support/ ?

File: test/cpp/end2end/end2end_test.cc [hunk 7]
```diff
@@ -98,6 +108,54 @@ bool CheckIsLocalhost(const grpc::string& addr) {
...
+  Status Process(const InputMetadata& auth_metadata, AuthContext* context,
+                 OutputMetadata* consumed_auth_metadata,
+                 OutputMetadata* response_metadata) GRPC_OVERRIDE {
+    EXPECT_TRUE(consumed_auth_metadata != nullptr);
+    EXPECT_TRUE(context != nullptr);
+    EXPECT_TRUE(response_metadata != nullptr);
+    auto auth_md = auth_metadata.find(GRPC_AUTHORIZATION_METADATA_KEY);
+    EXPECT_NE(auth_md, auth_metadata.end());
+    string_ref auth_md_value = auth_md->second;
+    if (auth_md_value.ends_with("Dr Jekyll")) {
```
Comment (yang-g) [inline test/cpp/end2end/end2end_test.cc]: use kGoodGuy here?

Comment (jboeuf) [inline test/cpp/end2end/end2end_test.cc]: good one. Thanks. Fixed.

**Comment by jboeuf** [comment]
Jenkins looks greenish. Interop is yellow and windows red but I don't if that's normal or not.

**Pull request merged**